File: 332. Reconstruct Itinerary.py
class Solution(object):
    def findItinerary(self, tickets):
        """
        :type tickets: List[List[str]]
        :rtype: List[str]
        """
        fromto = {}
        visited = {}
        self.n = len(tickets)
        for i in tickets:
            if i[0] not in fromto:
                fromto[i[0]] = []
                visited[i[0]] = []
            fromto[i[0]] += [i[1]]
            visited[i[0]] += [False]
        for key in fromto.keys():
            fromto[key] = sorted(fromto[key])
        self.res = []
        self.DFS(["JFK"], "JFK", fromto, visited)
        return self.res
    
    def DFS(self, path, start, fromto, visited):
        if len(path) == self.n + 1:
            self.res = path[:]
            return
        try:
            for i in xrange(len(fromto[start])):
                if visited[start][i] == False:
                    visited[start][i] = True
                    self.DFS(path + [fromto[start][i]], fromto[start][i], fromto, visited)
                    visited[start][i] = False
                    if len(self.res) > 0:
                        break
        except:
            return

File: 332. Reconstruct Itinerary.cpp
#include <unordered_map>
#include <stack>
class Solution {
public:
    vector<string> findItinerary(vector<pair<string, string>> tickets) {  
        vector<string> res;
        map<string, multiset<string>> fromto;  
        for(auto val:tickets)
            fromto[val.first].insert(val.second);
        stack <string> s;  
        s.push("JFK");
        
        while(!s.empty())
        {  
            string temp=s.top();  
            if(fromto.find(temp) != fromto.end() && fromto[temp].size() > 0)  
            {  
                s.push(*fromto[temp].begin());
                fromto[temp].erase(fromto[temp].begin());
            }
            else
            {
                res.insert(res.begin(), temp);
                s.pop();
            }
        }
        return res;  
    }  
};
